apiserver/watcher.go

```diff
@@ -86,46 +60,156 @@ type SrvAllWatcher struct {
+	deltas, err := aw.watcher.Next()
+	return params.AllWatcherNextResults{
+		Deltas: translate(newAllWatcherDeltaTranslaterV1(), deltas),
+	}, err
+}
+
+type allWatcherDeltaTranslaterV1 struct {
+	DeltaTranslater
+}
+
+func newAllWatcherDeltaTranslaterV1() DeltaTranslater {
+	return &allWatcherDeltaTranslaterV1{
+		newAllWatcherDeltaTranslater(),
+	}
+}
+
+func (aw allWatcherDeltaTranslaterV1) TranslateAction(info multiwatcher.EntityInfo) params.EntityInfo {
+	orig, ok := info.(*multiwatcher.ActionInfo)
+	if !ok {
+		logger.Criticalf("consistency error: %s", pretty.Sprint(info))
+		return nil
+	}
+	return &params.ActionInfo{
+		ModelUUID:  orig.ModelUUID,
+		Id:         orig.ID,
+		Receiver:   orig.Receiver,
+		Name:       orig.Name,
+		Parameters: orig.Parameters,
+		Status:     orig.Status,
+		Message:    orig.Message,
+		Results:    orig.Results,
+		Enqueued:   orig.Enqueued,
+		Started:    orig.Started,
+		Completed:  orig.Completed,
+	}
+}
+
+func newAllWatcher(context facade.Context) (*SrvAllWatcher, error) {
+	id := context.ID()
+	auth := context.Auth()
+	resources := context.Resources()
+
+	if !auth.AuthClient() {
+		// Note that we don't need to check specific permissions
+		// here, as the AllWatcher can only do anything if the
+		// watcher resource has already been created, so we can
+		// rely on the permission check there to ensure that
+		// this facade can't do anything it shouldn't be allowed
+		// to.
+		//
+		// This is useful because the AllWatcher is reused for
+		// both the WatchAll (requires model access rights) and
+		// the WatchAllModels (requring controller superuser
+		// rights) API calls.
+		return nil, apiservererrors.ErrPerm
+	}
+	watcher, ok := resources.Get(id).(multiwatcher.Watcher)
+	if !ok {
+		return nil, apiservererrors.ErrUnknownWatcher
+	}
+	return &SrvAllWatcher{
+		watcherCommon: newWatcherCommon(context),
+		watcher:       watcher,
+	}, nil
+}
+
+// NewAllWatcher returns a new API server endpoint for interacting
+// with a watcher created by the WatchAll and WatchAllModels API calls.
+func NewAllWatcher(context facade.Context) (facade.Facade, error) {
+	return newAllWatcher(context)
+}
+
 // Next will return the current state of everything on the first call
 // and subsequent calls will
 func (aw *SrvAllWatcher) Next() (params.AllWatcherNextResults, error) {
 	deltas, err := aw.watcher.Next()
 	return params.AllWatcherNextResults{
-		Deltas: aw.translate(deltas),
+		Deltas: translate(newAllWatcherDeltaTranslater(), deltas),
 	}, err
 }
 
-func (aw *SrvAllWatcher) translate(deltas []multiwatcher.Delta) []params.Delta {
+type allWatcherDeltaTranslater struct {
+	DeltaTranslater
+}
+
+func newAllWatcherDeltaTranslater() DeltaTranslater {
+	return &allWatcherDeltaTranslater{}
+}
+
+//go:generate go run github.com/golang/mock/mockgen -package mocks -destination mocks/deltatranslater_mock.go github.com/juju/juju/apiserver DeltaTranslater
```

Comment:
This should be in `package_test.go`. We've standardized there, so we don't have to look through all go files to locate them.